sandbox/libs/authn/docs/auth_flow.md

```diff
@@ -9,3 +9,35 @@ As requests are received by an OpenSearch node they need to be authenticated.  D
 The subject is associated for the lifetime of the request on that node.  If the request will be sent to other nodes within the cluster, the authentication information will be attached with a generated access token no matter the AuthenticationToken that was used to authenticate the user.  By including expiry on these tokens the exposure of the subjects permissions are greatly limited. 
 
 ![AccessToken for inter-node communication](https://user-images.githubusercontent.com/2754967/202580773-9b0ab15f-834c-45dc-9faf-48e6b832f85e.PNG)
+
+## Ensuring Authentication throughout OpenSearch
+
+As the OpenSearch service starts or in a running instance threads are created with the subject defaulting to Unauthenticated.  This is done to support backward compatibility as request authentication has never been a hard requirement of the service.
+
+**Why partition state by threads start?**
+Thread context is already widely used in OpenSearch to pass information through the transport layers.  This is not the most ideal state - and can be altered at a future time.  Many frameworks such as Shiro have static accessors for authentication information that is internally retrieved from a thread context.
+
+Any risks to this pattern?  Absolutely, naively usage of thread local storage for authentication information can leak into reused threads.  This is an elevation of privileges and can be hard to spot during testing, but triggerable in production scenarios.
```

Comment:
Naive question: When threads are freed back to the pool isn't that supposed to clear the thread context?

Reply:
>When threads are freed back to the pool isn't that supposed to clear the thread context?

This _should_ never happen, I would expected all kinds of unpredictable behavior would be seen in the product.  If we have any threads that operate like this they are 'stateful' in a fundamentally wrong ways, and would be a high-priority bug fix IMO.

Typically, if you have context on a thread that do something interesting, but is a pain to start up, you create that thread, but then have a `while(true)` loop that effectively manages tasks in that environment bound to that single thread.

Reply:
Note; we do have some strange 'sub' tasks patterns where tasks 'hint' which threadpool they should execute under.  As we improve this space the identity associated with the thread would determine where it could run, and we can ignore, then drop this concept completely.


https://github.com/opensearch-project/OpenSearch/blob/fb78f112a1e8939d649d79d3c4220ffbb43900c5/server/src/main/java/org/opensearch/threadpool/ThreadPool.java#L410-L414

Reply:
This is great as we won't run into re-usable Thread scenarios. Do you think if there is a way to kill a thread completely after its task is executed? If yes, we should estimate how expensive will it be to spin up new threads per task. Alternatively, can we clear "subject identifier" from the thread everytime a task is completed and add a new one based on authentication?

Reply:
The threading and object lifetime management systems are the job of the JVM.  We should advise no one implements tools to circumvent standard practices - a given if you ask me - and then we don't have to worry about any of this.

[Java Native Interface](https://www.ibm.com/support/pages/java-native-interface-jni) based code could do this, but we don't have code like that in OpenSearch's Core.